File: find-minimum-in-rotated-sorted-array/using-recursion.py
class Solution:
    def findMin(self, nums):
        def findPivotRecursive(nums, start, end):
            if start > end:
                return -1
            
            # pivot in start
            if start > 1 and nums[start] < nums[start-1]:
                return start
            
            mid = (start + end) // 2

            # pivot in middle
            if mid > 0 and nums[mid] < nums[mid-1]:
                return mid

            # pivot in left half
            if nums[start] > nums[mid]:
                return findPivotRecursive(nums, start, mid-1)
            # pivot in right half
            elif nums[mid] > nums[end]:
                return findPivotRecursive(nums, mid+1, end)
            # pivot not in this range
            else:
                return -1

        if len(nums) == 0:
            return 0
            
        pivot = findPivotRecursive(nums, 0, len(nums)-1)
        if pivot < 0:
            return nums[0]
        else:
            return nums[pivot]

File: convert-sorted-array-to-binary-search-tree/using-recursion.cpp
#include<vector>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        if (nums.size() == 0) {
            return nullptr;
        }
        return usingRecursion(nums, 0, nums.size()-1);
    }
private:
    TreeNode* usingRecursion(vector<int> &nums, int start, int end) {
        if (start > end) {
            return nullptr;
        }
        int mid = (start+end)/2;
        TreeNode *root = new TreeNode(nums[mid]);
        root->left = usingRecursion(nums, start, mid-1);
        root->right = usingRecursion(nums, mid+1, end);
        return root;
    }
};